### PR TITLE
[RTM] Recursively replace insert tags

### PR DESCRIPTION
**Pull request by leofeyer**
This PR implements contao/core#8172. It also allows us to remove the "nested insert tags" routines.

### CODE REVIEW

**Comment by leofeyer** [comment]
This is RTM to me. @Toflar What about you?

File: src/Resources/contao/library/Contao/InsertTags.php [hunk 1]
```diff
@@ -52,15 +52,20 @@ public function replace($strBuffer, $blnCache=true)
 			return \StringUtil::restoreBasicEntities($strBuffer);
 		}
 
-		$tags = preg_split('/{{(([^{}]*|(?R))*)}}/', $strBuffer, -1, PREG_SPLIT_DELIM_CAPTURE);
+		$tags = preg_split('/{{([^{}]+)}}/', $strBuffer, -1, PREG_SPLIT_DELIM_CAPTURE);
+
+		if (count($tags) < 2)
+		{
+			return \StringUtil::restoreBasicEntities($strBuffer);
```
Comment (leofeyer) [inline src/Resources/contao/library/Contao/InsertTags.php]: @Toflar This new return statement actually prevents endless loops now, so we do no longer need the nesting level counter.

Comment (Toflar) [inline src/Resources/contao/library/Contao/InsertTags.php]: Ah true, why did I not think about this^^

**Comment by Toflar** [comment]
Looks good, I think. Maybe @fritzmg can try wether this code solves his problem. Moreover, are there maybe other places where we can now skip replacing insert tags because they are recursive now?

**Comment by leofeyer** [comment]
> Moreover, are there maybe other places where we can now skip replacing insert tags because they are recursive now?

I bet there are. Searching …

**Comment by leofeyer** [comment]
Unexpectedly, I did not find any places where the `replaceInsertTags()` call can be omitted.

**Comment by fritzmg** [comment]
> Maybe @fritzmg can try wether this code solves his problem.

Well actually I didn't experience a problem (directly) - @aschempp had the following problem originally: https://github.com/contao/core/issues/7097#issuecomment-126189860

I reinvestigated this because I was curious about whether insert tags are not replaced recursively in the first place and because it would have solved some other problems that were reported in the past by other users.

**Comment by Toflar** [comment]
Ah, true. Then imho this should be fine.

**Comment by leofeyer** [comment]
Merged in c05c02a0dc79015f8be1f7be19c6608cdb3af289.

**Comment by leofeyer** [comment]
Unfortunately, it does not work if there are non-cacheable insert tags. `count($tags)` will be greater than 2 in this case and we are back in the endless loop. 😞

**Comment by leofeyer** [comment]
Luckily, it was easy to fix: ab4434da67136c15ed28cb02cca869249199dcb7

**Comment by aschempp** [comment]
Doesn’t that mean we will not recursive replace insert tags if the content is not cached?

**Comment by fritzmg** [comment]
Hm, consider the following example:

Some content somewhere:
```
{{insert_content::123}}
```

Content of content element 123:
```
{{insert_content::124}}
```

When the page is generated for the cache, `{{insert_content::123}}` will be replaced by `{{insert_content::124}}`. But that insert tag will not be replaced anymore, if I understand the current code correclty, thus it will land in the cache and it also will not be replaced for the frontend output during cache generation, I think?

I think the only way to properly implement recursive insert tag replacing while allowing cacheable insert tags is with the checksum method that @Toflar originally thought up [here](https://github.com/contao/core/issues/8172#issuecomment-167105203).

**Comment by aschempp** [comment]
As I pointed out in https://github.com/contao/core/issues/8172#issuecomment-185710174, we don't even need a checksum. Haste simply compares if the execution of `replaceInsertTag` gives the same result as before and if so then stops the loop…

**Comment by leofeyer** [comment]
But this would mean an additional wrapper method, wouldn't it?

**Comment by aschempp** [comment]
Why? Just check original content passed to the method with the return content before calling again?

**Comment by leofeyer** [comment]
Fixed in 824b012eaefbabcba9cac59881bf27f6bfcdffca.

**Comment by aschempp** [comment]
I would still do the `strpos($strBuffer, '{{') !== false` to prevent the whole replace thing to run again if not necessary.

**Comment by leofeyer** [comment]
Re-added in 976a568834b8b6f065caf3ae643497cddefda480.

**Pull request merged**